File: Assignment/CODING/Cau2/Cau2.java
import java.util.*;
import java.io.*;
import java.util.stream.Collectors;

class FunctionalDependency {
    Set<String> lhs;
    Set<String> rhs;

    public FunctionalDependency(Set<String> lhs, Set<String> rhs) {
        this.lhs = lhs;
        this.rhs = rhs;
    }

    @Override
    public String toString() {
        return lhs.stream()
                .sorted()
                .collect(Collectors.joining()) + " -> "
                + rhs.stream()
                        .sorted()
                        .collect(Collectors.joining());
    }
}

class FDSet extends HashSet<FunctionalDependency> {
    public Set<String> setOfAttributesInLeftHandSide() {
        Set<String> result = new HashSet<>();
        this.forEach(fd -> result.addAll(fd.lhs));
        return result;
    }

    public Set<String> setOfAttributesInRightHandSide() {
        Set<String> result = new HashSet<>(this.setOfAttributesInLeftHandSide());
        result.removeAll(this.setOfAttributesInLeftHandSide());
        return result;
    }

    public Set<String> closureOfAttributeSet(Set<String> attributes) {
        Set<String> result = new HashSet<>(attributes);
        boolean stillChange = true;
        while (stillChange) {
            stillChange = false;
            for (FunctionalDependency fd : this) {
                if (fd.lhs.stream().allMatch(result::contains) && !fd.rhs.stream().allMatch(result::contains)) {
                    stillChange = true;
                    result.addAll(fd.rhs);
                }
            }
        }
        return result;
    }

    public List<String> keys(Set<String> attributes) {
        Set<String> N = new HashSet<>(attributes);
        N.removeAll(setOfAttributesInRightHandSide());
        if (convertSetToString(closureOfAttributeSet(N)).equals(convertSetToString(attributes))) {
            return new ArrayList<>(N);
        }
        Set<String> D = new HashSet<>(setOfAttributesInRightHandSide());
        D.removeAll(setOfAttributesInLeftHandSide());
        Set<String> L = new HashSet<>(attributes);
        L.removeAll(N);
        L.removeAll(D);

        // Xây dựng tập hợp con của L
        List<Set<String>> powerSet = new ArrayList<>();
        powerSet.add(new HashSet<>()); // Thêm tập hợp rỗng làm bước ban đầu

        for (String attr : L) {
            List<Set<String>> newSets = new ArrayList<>();
            for (Set<String> subset : powerSet) {
                Set<String> newSubset = new HashSet<>(subset);
                newSubset.add(attr);
                newSets.add(newSubset);
            }
            powerSet.addAll(newSets);
        }

        List<String> result = new ArrayList<>();
        for (Set<String> Li : powerSet) {
            Set<String> X = new HashSet<>(N);
            X.addAll(Li);
            Set<String> closureX = closureOfAttributeSet(X);
            if (sortString(convertSetToString(closureX)).equals(sortString(convertSetToString(attributes)))) {
                result.add(convertSetToString(X));
            }
        }
        return result;
    }

    // Để tạo chuỗi từ một tập hợp
    public static String convertSetToString(Set<String> set) {
        return set.stream()
                .sorted()
                .collect(Collectors.joining());
    }

    // Để sắp xếp chuỗi
    public static String sortString(String str) {
        return str.chars()
                .sorted()
                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
                .toString();
    }

    public static int countLinesInFile(String filePath) throws IOException {
        int count = 0;
        try (BufferedReader myFile = new BufferedReader(new FileReader(filePath))) {
            while (myFile.readLine() != null) {
                count++;
            }
        }
        return count;
    }
}

class Relation {
    String name;
    String attributes;
    FDSet fdSet;
}

class Main {
    public static List<Relation> loadFile(String filePath) throws IOException {
        List<Relation> relations = new ArrayList<>();
        try (BufferedReader myInput = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = myInput.readLine()) != null) {
                String[] data = line.split(" ");
                Relation relation = new Relation();
                relation.name = data[0];
                relation.attributes = data[1];
                relation.fdSet = new FDSet();
                String[] fdSet = data[2].split(";");
                for (String fd : fdSet) {
                    String[] fdData = fd.split("_");
                    relation.fdSet.add(new FunctionalDependency(
                            Arrays.stream(fdData[0].split(",")).collect(Collectors.toSet()),
                            Arrays.stream(fdData[1].split(",")).collect(Collectors.toSet())));
                }
                relations.add(relation);
            }
        }
        return relations;
    }

    public static void writeToFile(String filePath, List<Relation> relations, String X) throws IOException {
        try (BufferedWriter myOutput = new BufferedWriter(new FileWriter(filePath))) {
            for (Relation relation : relations) {
                myOutput.write(convertSetToString(
                        relation.fdSet.closureOfAttributeSet(new HashSet<>(Arrays.asList(X.split(","))))) + " "
                        + String.join(",",
                                relation.fdSet.keys(new HashSet<>(Arrays.asList(relation.attributes.split(","))))));
                myOutput.newLine();
            }
        }
    }

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter attribute set X: ");
        String X = scanner.nextLine();
        List<Relation> relations = loadFile("input.txt");
        writeToFile("output.txt", relations, X);
        scanner.close();
    }

    public static String convertSetToString(Set<String> set) {
        return set.stream()
                .sorted()
                .collect(Collectors.joining());
    }
}

File: Assignment/CODING/Cau2/Cau2.py
import itertools

# Hàm chuyển đổi một tập hợp thành một chuỗi
def convert_set_to_string(attribute_set):
    '''Lặp qua các phần tử trong tập hợp và ghép chúng lại thành một chuỗi'''
    return "".join(attribute for attribute in attribute_set)

#  Hàm chuyển đổi một chuỗi thành một tập hợp
def convert_string_to_set(string):
    '''Tách chuỗi theo dấu phẩy và tạo một tập hợp từ các phần tử tách được'''
    attributes = string.split(",")
    return set(attributes)

# Lớp phụ thuộc hàm
class FunctionalDependency:
    def __init__(self, LeftAttribute, RightAttributes):
        self.LeftAttribute = LeftAttribute
        self.RightAttributes = RightAttributes

    def __str__(self):
        return "{} -> {}".format(convert_set_to_string(self.LeftAttribute), convert_set_to_string(self.RightAttributes))

# Lớp chứa các phụ thuộc hàm
class FDSet(set):
    # Hàm cập nhật phụ thuộc hàm ở bên trái 
    def UL(self):
        resLeftAttributet = set()
        for fd in self:
            resLeftAttributet.update(fd.LeftAttribute)
        return resLeftAttributet

    # Hàm cập nhật phụ thuộc hàm ở bên phải
    def UR(self):
        resLeftAttributet = set()
        for fd in self:
            resLeftAttributet.update(fd.RightAttributes)
        return resLeftAttributet

    # Hàm tính bao đóng
    # def FindClose(self, attributes):
    #     X = set(attributes)
    #     still_change = True
    #     while still_change:
    #         still_change = False
    #         '''Duyệt qua tất cả phụ thuộc hàm'''
    #         for fd in self:
    #             '''Nếu các thuộc tính bên trái của phụ thuộc hàm là tập con của tập X
    #             và (các thuộc tính bên phải chưa có trong X) thì cập nhật các thuộc tính bên phải đó vào X'''
    #             if fd.LeftAttribute.issubset(X) and not fd.RightAttributes.issubset(X):
    #                 still_change = True
    #                 X.update(fd.RightAttributes)
    #     return X
    def FindClose(self, attributes):
        X = set(attributes)
        prev_length = 0

        while True:
            for fd in self:
                if fd.LeftAttribute.issubset(X) and not fd.RightAttributes.issubset(X):
                    X.update(fd.RightAttributes)

            # Kiểm tra xem tập X có thêm được bất kỳ thuộc tính mới nào không
            if len(X) == prev_length:
                break
            prev_length = len(X)

        return X

    
    

    # Hàm tìm khóa
    def Findkeys(self, attributes):
        '''N = U - UR'''
        N = attributes.difference(self.UR())
        '''Nếu N+ = U thì N là khóa duy nhất'''
        if self.FindClose(N) == attributes:
            return [convert_set_to_string(N)]
        '''D = UR - UL'''
        D = self.UR().difference(
            self.UL())
        '''L = U - N union D'''
        L = attributes.difference(N.union(D))
        Keys = [] # Khởi tạo tập Key chứa các khóa
        power_set_L = N
        num_attributes = L
        '''Duyệt Li gồm 1 thuộc tính'''
        limit = len(num_attributes) // 2 + 1

        for i in range(1, len(L) + 1):
            subsets = itertools.combinations(attributes, i)
            for subset in subsets:
                for key in Keys:
                        if key.issubset(subset):
                            continue
                        subset_closure = self.FindClosure(set(subset))
                        if subset_closure == attributes:
                            Keys.append(subset)
                    
            # Check if the number of attributes in the subset exceeds the limit
            if i >= limit:
                break

            return Keys       
        
        # resLeftAttributet = []
        # Key = [] # Khởi tạo tập Key chứa các khóa
        
        # for Li in power_set_L:
        #     is_key = False
        #     Li = set(Li)
        #     '''Duyệt Li qua các khóa trong tập Key nếu là tập con của khóa thì bỏ qua'''
        #     for element in Key:
        #         if element.issubset(Li):
        #             is_key = True
        #     if is_key:
        #         continue
        #     '''X = N uinion Li'''
        #     X = N.union(Li)
        #     '''Nếu X+ = U thì '''
        #     if self.FindClose(X) == attributes:
        #         resLeftAttributet.append(convert_set_to_string(X))
        #         Key.append(Li)
        
        return Keys

# Class to represent a Relation
class Relation:
    def __init__(self, name="", attributes=""):
        self.name = name
        self.attributes = attributes
        self.fd_set = FDSet()

# Function to count lines in a file
def count_lines_in_file(file_path):
    count = 0
    with open(file_path, "r") as my_file:
        lines = my_file.readlines()
        count = len(lines)
    return count

# Function to load data from a file into relations
def load_file(file_path):
    relations = []
    with open(file_path, "r") as my_input:
        lines = my_input.readlines()
        for line in lines:
            data = line.rstrip('\n').split(" ")
            relation = Relation(data[0], data[1])
            fd_set = data[2].split(";")
            for fd in fd_set:
                relation.fd_set.add(FunctionalDependency(
                    convert_string_to_set(fd.split("_")[0]),
                    convert_string_to_set(fd.split("_")[1])
                ))
            relations.append(relation)
    return relations

# Function to write data to a file
def write_to_file(file_path, relations, X):
    with open(file_path, "w") as my_output:
        for relation in relations:
            my_output.writelines(
                convert_set_to_string(relation.fd_set.FindClose(X)) +
                " " +
                ','.join(relation.fd_set.Findkeys(convert_string_to_set(relation.attributes)))
            )
            my_output.write("\n")


# RunCode
X = input("Enter attribute set X: ").rstrip('\n')
relations = load_file("input2.txt")
write_to_file("output2.txt", relations, X)


